packages/relay/src/index.ts

```diff
@@ -116,50 +128,51 @@ class Relay implements IRelay {
       };
     }
 
-    return this.gossipSub.publish(this.pubSubTopic, msg);
+    return this.gossipSub.publish(pubSubTopic, msg);
   }
 
-  /**
-   * Add an observer and associated Decoder to process incoming messages on a given content topic.
-   *
-   * @returns Function to delete the observer
-   */
   public subscribe<T extends IDecodedMessage>(
     decoders: IDecoder<T> | IDecoder<T>[],
     callback: Callback<T>
   ): () => void {
-    const contentTopicToObservers = Array.isArray(decoders)
-      ? toObservers(decoders, callback)
-      : toObservers([decoders], callback);
-
-    for (const contentTopic of contentTopicToObservers.keys()) {
-      const currObservers = this.observers.get(contentTopic) || new Set();
-      const newObservers =
-        contentTopicToObservers.get(contentTopic) || new Set();
-
-      this.observers.set(contentTopic, union(currObservers, newObservers));
+    const observers: Array<[PubSubTopic, Observer<T>]> = [];
+
+    for (const decoder of Array.isArray(decoders) ? decoders : [decoders]) {
+      const pubSubTopic = decoder.pubSubTopic;
+      const ctObs: Map<ContentTopic, Set<Observer<T>>> = this.observers.get(
+        pubSubTopic
+      ) ?? new Map();
+      const _obs = ctObs.get(decoder.contentTopic) ?? new Set();
+      const observer = { pubSubTopic, decoder, callback };
+      _obs.add(observer);
```

Comment:
I wonder if `pushOrInitMap` could be use here. Probably not as it's a `Set` and not an `Array`.